### PR TITLE
CORCI-953 build: Update HDF5 to build with openmpi

### PR DESCRIPTION
**Pull request by mjean308**
Signed-off-by: Maureen Jean <maureen.jean@intel.com>

### CODE REVIEW

**Review by phender — APPROVED** [review]
LGTM

**Review by brianjmurrell — COMMENTED** [review]
You should do a daos master run with:

```
PR-repos: hdf5@PR-7
```

in the commit message just to be sure landing this does not cause daos master to start breaking.  Please post the link to the PR here when you do.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
Only -1 here to prevent accidental landing before we get the daos master test run complete.

You can skip the Functional hardware tests in that run, just to make it go more quickly.  The VM run will test everything we need to see.

**Comment by mjean308** [comment]
The daos PR that is using the rpms generated from this PR is:

https://github.com/daos-stack/daos/pull/3095

**Comment by brianjmurrell** [comment]
> The daos PR that is using the rpms generated from this PR is:
> 
> [daos-stack/daos#3095](https://github.com/daos-stack/daos/pull/3095)

So this is not what I was requesting prior.  While this is the next step after this PR lands,you need to do a test run with daos master that does not alter *anything* except add this PR via the `PR-repos: hdf5@PR-7` commit pragma.  That test run tells us what PRs and master will do (which is hopefully continue to run and pass tests) in the window of time after this lands, but before daos-stack/daos#3095 lands and everyone rebases to pick up the changes.

**Comment by mjean308** [comment]
I will do  a run with only The rpms and no other changes.  I expect that it will fail because this PR is no longer building an rpm called hdf5-tests, instead it is now hdf5-mpich-tests and hdf5-openmpi-tests.    Should I leave the mpich built tests as is?   ie. hdf5-tests?   The hdf5 tests in the daos PR did pass and it did use the newly created hdf5-mpich-tests rpms.

**Comment by brianjmurrell** [comment]
> I expect that it will fail because this PR is no longer building an rpm called hdf5-tests, instead it is now hdf5-mpich-tests and hdf5-openmpi-tests.

But RPMs of the existing names still exist in the repo and so that test run should use those, even when the ones from this PR are present (by way of the `PR-repos: hdf5@PR-7`.  Seeing that a master run can pass even when the new RPMs are also available is actually the entire point of the test run.  We cannot allow the landing of this PR to make any other PRs that are in flight fail because of the renaming.

Ahhh.  But this brings up an issue...

> Should I leave the mpich built tests as is? ie. hdf5-tests? The hdf5 tests in the daos PR did pass and it did use the newly created hdf5-mpich-tests rpms.

Since this PR renames hdf5-tests (which was build with mpich) to hdf5-mpich-tests, you need to have that new package `Obsoletes: %{name)-tests < %{version}-%{release}` so that somebody installing this new package over top of the old one properly replaces it.

File: hdf5.spec
Comment (brianjmurrell) [inline hdf5.spec:175]:
```suggestion
%package mpich-tests
Obsoletes: %{name}-tests < %{version}-%{release}
```

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: hdf5.spec
Comment (brianjmurrell) [inline hdf5.spec]: Why is this necessary?  This new packaging is named differently than the old packaging, so existing PRs and master should use the existing/old packages and these new packages, with the new paths won't be used until daos is updated to use the new names.

File: hdf5.spec
Comment (brianjmurrell) [inline hdf5.spec]: Ditto.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

**Comment by mjean308** [comment]
Brian,

The DAOS PR, ( which uses the PR-repos: hdf5@PR-7 and makes NO changes to DAOS tests)  PR-3095 has passed and then I also created PR-3117 which includes the Jenkins file and hdf5 yaml updates to allow the tests to use the hdf5-mpich lib.  That PR also passed.  I verified that each PR is using the correct hdf5 rpm packages.    PR-3095 used the 1.10.5-8 (existing version) and PR-3117 used the   1.10.5-9 version.    When I update the hdf5 rpm to 1.12; I can add in the "obsolete ".     Is this ready to merge now?

File: hdf5.spec
Comment (brianjmurrell) [inline hdf5.spec]: I'm on the fence about this.  EL7 does have both an openmpi and openmpi3 package which are two different major versions of openmpi.  While I can't see us ever wanting to have openmpi (as opposed to openmpi3) support, overloading openmpi to really mean openmpi3 here gives me pause.

I noticed you did use openmpi3 as the variant name in a previous iteration of this patch.  Can I ask why you backed away from that?

As just one randomly chosen example:

```
$ dnf search lammps
========================= Name Exactly Matched: lammps =========================
lammps.x86_64 : Molecular Dynamics Simulator
======================== Name & Summary Matched: lammps ========================
lammps-data.noarch : Data files for LAMMPS
python36-lammps.x86_64 : LAMMPS Python interface
lammps-devel.x86_64 : Development libraries for LAMMPS
lammps-headers.x86_64 : Development headers for LAMMPS
lammps-mpich.x86_64 : LAMMPS MPICH binaries and libraries
lammps-openmpi.x86_64 : LAMMPS Open MPI binaries and libraries
lammps-openmpi3.x86_64 : LAMMPS Open MPI 3 binaries and libraries
lammps-mpich-devel.x86_64 : Development libraries for MPICH LAMMPS
lammps-openmpi-devel.x86_64 : Development libraries for Open MPI LAMMPS
lammps-openmpi3-devel.x86_64 : Development libraries for Open MPI 3 LAMMPS
```

File: hdf5.spec
Comment (brianjmurrell) [inline hdf5.spec]: I wonder if doing this distro conditional here (and elsewhere) is the right way of doing this or if the names in the list `%{mpi_list}` should be conditionalized per distro.

File: hdf5.spec
Comment (mjean308) [reply hdf5.spec]: The changes from openmpi3 to openmpi was based on feedback from a previous review.   The change was requested due to  how mpich was referenced (mpich and mpich2).      I think that the name  should have openmpi3*.rpm (version 3.x) and if we decide we need it, openmpi (for version 1.x) in the package name.    I will update the package names.

Comment (mjean308) [reply hdf5.spec]: mpi_list will now have openmpi and openmpi3 and the above conditional is no longer needed.

File: hdf5.spec
Comment (brianjmurrell) [inline hdf5.spec]: Yeah, seeing this all come back, I think the SUSE/RH conditionals should be around the %{mpi_list} construction, not it's evaluation.

Or alternatively (and better yet), put all of this into a macro called `load_mpi` or something so that the execution locations, like this one are much cleaner with just: `load_mpi $mpi` and let the macro do all the work of figuring out the command and correct name to load, etc.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: hdf5.spec
Comment (mjean308) [reply hdf5.spec]: Phil had recommended the same macro but I could not get it to pass.  The changes we made were:

if (0%{?suse_version} >= 1500)	
%global load_module() (if [ "%{1}" == "openmpi3" ]; then module load gnu-openmpi; else module load gnu-%{1}; fi)	
%else	
%global load_module() (module load mpi/%{1}-%{_arch})	
%endif

and it would not build with mpich; so I reverted the changes and it now passes.   Does the macro def above look correct?

Comment (brianjmurrell) [reply hdf5.spec]: So let's resurrect the macro and figure out why it fails.

**Review by phender — APPROVED** [review]

File: hdf5.spec
Comment (phender) [reply hdf5.spec]: wrapping the macro body in `()` was causing the `module load` to be run in a sub-shell, preventing the environment from being set correctly.

**Review by brianjmurrell — APPROVED** [review]
@mjean308 Do you want me to hold off on landing this or are you confident in your previous tests of vanilla master with PR-repos: hdf5@PR-7 are still valid with the new changes here?  Or do you want to run that again to be sure landing this doesn't break master?  I'm fine either way.  It's up to you.

**Comment by mjean308** [comment]
@brianjmurrell I am running the two daos PRs (3095 and 3117)  with this PR now.  As soon as they pass we can land this.

**Pull request merged**
